File: src/graalvm/grpc/grpc-java/src/main/java/benchmark/DatabaseService.java
package benchmark;

import benchmark.Benchmark.*;
import com.zaxxer.hikari.HikariConfig;
import com.zaxxer.hikari.HikariDataSource;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.sql.*;
import java.util.ArrayList;
import java.util.List;

public class DatabaseService {

    private static final Logger logger = LoggerFactory.getLogger(DatabaseService.class);
    private final HikariDataSource dataSource;

    public DatabaseService() {
        String dbUrl = System.getenv().getOrDefault("DATABASE_URL", "jdbc:postgresql://localhost:5432/benchmark");
        String dbUser = System.getenv().getOrDefault("DB_USER", "benchmark");
        String dbPassword = System.getenv().getOrDefault("DB_PASSWORD", "benchmark");

        HikariConfig config = new HikariConfig();
        config.setJdbcUrl(dbUrl);
        config.setUsername(dbUser);
        config.setPassword(dbPassword);
        config.setMaximumPoolSize(25);
        config.setMinimumIdle(5);
        config.setDriverClassName("org.postgresql.Driver");

        this.dataSource = new HikariDataSource(config);
        logger.info("Database connection pool initialized");
    }

    public boolean isConnected() {
        try (Connection conn = dataSource.getConnection()) {
            return conn.isValid(5);
        } catch (SQLException e) {
            return false;
        }
    }

    public UserResponse getUser(int id) throws SQLException {
        String sql = "SELECT id, email, first_name, last_name, age, created_at FROM users WHERE id = ?";

        try (Connection conn = dataSource.getConnection();
             PreparedStatement stmt = conn.prepareStatement(sql)) {

            stmt.setInt(1, id);

            try (ResultSet rs = stmt.executeQuery()) {
                if (rs.next()) {
                    return UserResponse.newBuilder()
                            .setId(rs.getInt("id"))
                            .setEmail(rs.getString("email"))
                            .setFirstName(rs.getString("first_name"))
                            .setLastName(rs.getString("last_name"))
                            .setAge(rs.getInt("age"))
                            .setCreatedAt(rs.getTimestamp("created_at").toInstant().toString())
                            .build();
                }
                return UserResponse.getDefaultInstance();
            }
        }
    }

    public ComplexOrdersResponse getComplexOrders(int days) throws SQLException {
        String sql = """
                SELECT
                    u.id AS user_id,
                    u.first_name || ' ' || u.last_name AS user_name,
                    COUNT(o.id) AS total_orders,
                    COALESCE(SUM(o.total_amount), 0) AS total_value,
                    COALESCE(AVG(o.total_amount), 0) AS average_order_value
                FROM users u
                LEFT JOIN orders o ON u.id = o.user_id
                    AND o.created_at >= NOW() - INTERVAL '1 day' * ?
                GROUP BY u.id, u.first_name, u.last_name
                ORDER BY total_value DESC
                LIMIT 100
                """;

        try (Connection conn = dataSource.getConnection();
             PreparedStatement stmt = conn.prepareStatement(sql)) {

            stmt.setInt(1, days);

            try (ResultSet rs = stmt.executeQuery()) {
                ComplexOrdersResponse.Builder builder = ComplexOrdersResponse.newBuilder();
                builder.setPeriodDays(days);

                List<UserOrderStats> statsList = new ArrayList<>();
                while (rs.next()) {
                    statsList.add(UserOrderStats.newBuilder()
                            .setUserId(rs.getInt("user_id"))
                            .setUserName(rs.getString("user_name"))
                            .setTotalOrders(rs.getInt("total_orders"))
                            .setTotalValue(rs.getDouble("total_value"))
                            .setAverageOrderValue(rs.getDouble("average_order_value"))
                            .build());
                }

                builder.setTotalUsers(statsList.size());
                builder.addAllData(statsList);
                return builder.build();
            }
        }
    }

    public void close() {
        if (dataSource != null && !dataSource.isClosed()) {
            dataSource.close();
            logger.info("Database connection pool closed");
        }
    }
}

File: src/csharp/grpc/grpc-dotnet/Services/DatabaseService.cs
using Npgsql;
using Benchmark;

namespace GrpcDotnet.Services;

public class DatabaseService
{
    private readonly string _connectionString;

    public DatabaseService(IConfiguration configuration)
    {
        _connectionString = configuration.GetValue<string>("ConnectionStrings:PostgreSQL")
            ?? Environment.GetEnvironmentVariable("DATABASE_URL")
            ?? "Host=localhost;Port=5432;Database=benchmark;Username=benchmark;Password=benchmark";
    }

    public async Task<UserResponse> GetUserAsync(int id)
    {
        await using var connection = new NpgsqlConnection(_connectionString);
        await connection.OpenAsync();

        await using var cmd = new NpgsqlCommand(
            "SELECT id, email, first_name, last_name, age, created_at FROM users WHERE id = @id", connection);
        cmd.Parameters.AddWithValue("id", id);

        await using var reader = await cmd.ExecuteReaderAsync();
        if (await reader.ReadAsync())
        {
            return new UserResponse
            {
                Id = reader.GetInt32(0),
                Email = reader.GetString(1),
                FirstName = reader.GetString(2),
                LastName = reader.GetString(3),
                Age = reader.GetInt32(4),
                CreatedAt = reader.GetDateTime(5).ToString("o")
            };
        }

        return new UserResponse();
    }

    public async Task<ComplexOrdersResponse> GetComplexOrdersAsync(int days)
    {
        await using var connection = new NpgsqlConnection(_connectionString);
        await connection.OpenAsync();

        var sql = @"
            SELECT
                u.id AS user_id,
                u.first_name || ' ' || u.last_name AS user_name,
                COUNT(o.id) AS total_orders,
                COALESCE(SUM(o.total_amount), 0) AS total_value,
                COALESCE(AVG(o.total_amount), 0) AS average_order_value
            FROM users u
            LEFT JOIN orders o ON u.id = o.user_id AND o.created_at >= NOW() - INTERVAL '1 day' * @days
            GROUP BY u.id, u.first_name, u.last_name
            ORDER BY total_value DESC
            LIMIT 100";

        await using var cmd = new NpgsqlCommand(sql, connection);
        cmd.Parameters.AddWithValue("days", days);

        var stats = new List<UserOrderStats>();
        await using var reader = await cmd.ExecuteReaderAsync();
        while (await reader.ReadAsync())
        {
            stats.Add(new UserOrderStats
            {
                UserId = reader.GetInt32(0),
                UserName = reader.GetString(1),
                TotalOrders = reader.GetInt32(2),
                TotalValue = reader.GetDouble(3),
                AverageOrderValue = reader.GetDouble(4)
            });
        }

        return new ComplexOrdersResponse
        {
            PeriodDays = days,
            TotalUsers = stats.Count,
            Data = { stats }
        };
    }
}
